lib/improver/spotdata/extract_data.py

```diff
@@ -319,56 +454,120 @@ def model_level_temperature_lapse_rate(self, cube, sites, neighbours,
                              'suitable.'.format(self.method, cube.name()))
 
         kappa = R_DRY_AIR/CP_DRY_AIR
 
+        z_lower, = lower_pressure.coord('height').points
+        z_upper, = upper_pressure.coord('height').points
+        dz_model_levels = z_upper - z_lower
+
         data = np.empty(shape=(len(sites)))
         for i_site in range(len(sites)):
-
             i, j, dz = (neighbours['i'][i_site], neighbours['j'][i_site],
                         neighbours['dz'][i_site])
 
-            # Use neighbour grid point value if vertical displacement == 0.
-            if isclose(dz, 0.):
+            # Use neighbour grid point value if dz < dz_tolerance.
+            if abs(dz) < dz_tolerance:
                 data[i_site] = cube.data[i, j]
                 continue
 
-            t_upper = temperature_on_height_levels.data[i, j]
-            p_upper = pressure_on_height_levels.data[i, j]
+            # Temperatures at surface, lower model level and upper model level.
             t_surface = cube.data[i, j]
-            p_surface = surface_pressure.data[i, j]
+            t_lower = lower_temperature.data[i, j]
+            t_upper = upper_temperature.data[i, j]
 
-            p_grad = (p_upper - p_surface)/50.
-            p_site = p_surface + p_grad*dz
-
-            theta_upper = t_upper*(1.0E5/p_upper)**kappa
-            theta_surface = t_surface*(1.0E5/p_surface)**kappa
-            dthetadz = (theta_upper - theta_surface)/50.
-
-            if abs(dz) < 1.:
-                t1p5 = t_surface
+            # Pressures at surface, lower model level and upper model level.
+            p_surface = surface_pressure.data[i, j]
+            p_lower = lower_pressure.data[i, j]
+            p_upper = upper_pressure.data[i, j]
+
+            # Potential temperature at surface, lower model level and upper
+            # model level.
+            theta_surface = t_surface*(p_ref/p_surface)**kappa
+            theta_lower = t_lower*(p_ref/p_lower)**kappa
+            theta_upper = t_upper*(p_ref/p_upper)**kappa
+
+            # Enforce a maximum vertical displacement to which temperatures
+            # will be adjusted using a lapse rate. This is to prevent excessive
+            # changes based on what is unlikely to be a constant gradient
+            # in potential temperature.
+            dz = min(abs(dz), dz_max_adjustment)*np.sign(dz)
+
+            # Calculate potential temperature gradient using levels away from
+            # surface.
+            dthetadz = (theta_upper - theta_lower)/dz_model_levels
+            if dthetadz <= dthetadz_threshold:
+                # If the potential temperature gradient is below the defined
+                # dthetadz_threshold, use the non-surface levels.
+                dz_from_model_level = dz - z_lower
+                p_grad = (p_upper - p_lower)/dz_model_levels
+                p_site = p_lower + p_grad*dz_from_model_level
```

Comment:
I think that you're assuming here that the change in pressure is linear with the change in height. It might be worth adding this assumption, either in the docstring, or as a comment.